File: Algorithms/src/3289. The Two Sneaky Numbers of Digitville/Code/solution.py
from typing import List

class Solution:
    def getSneakyNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums) - 2                   # original range size
        seen = [False] * n                  # boolean marks for 0..n-1
        res = []
        for x in nums:
            if seen[x]:
                res.append(x)               # found a duplicate
                if len(res) == 2:
                    break                   # early exit
            else:
                seen[x] = True              # mark seen
        return res

File: Algorithms/src/3289. The Two Sneaky Numbers of Digitville/Code/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> getSneakyNumbers(vector<int>& nums) {
        int n = (int)nums.size() - 2;          // original range size
        vector<char> seen(n, 0);              // boolean marks (char to save a bit)
        vector<int> res;
        res.reserve(2);
        for (int x : nums) {
            if (seen[x]) {
                res.push_back(x);             // found a repeated number
            } else {
                seen[x] = 1;                 // mark seen
            }
            if ((int)res.size() == 2) break; // early exit once we have both
        }
        return res;
    }
};
